ml/backend/ggml/ggml.go

```diff
@@ -312,24 +320,53 @@ func New(r *os.File, params ml.BackendParams) (ml.Backend, error) {
 					return fmt.Errorf("unassigned tensor: %s", t.Name)
 				}
 
-				bts := C.malloc(C.size_t(t.Size()))
-				if bts == nil {
-					return errors.New("failed to allocate tensor buffer")
+				tts[i] = tt
+			}
+
+			bsr := bufio.NewReader(io.NewSectionReader(sr, int64(t.Offset), int64(t.Size())))
+			bts := make([]byte, 32*format.KibiByte)
+
+			var s uint64
+			for s < t.Size() {
+				n, err := io.ReadFull(bsr, bts[:min(len(bts), int(t.Size()-s))])
+				if err != nil {
+					return err
 				}
-				defer C.free(bts)
 
-				buf := unsafe.Slice((*byte)(bts), t.Size())
-				n, err := io.ReadFull(io.NewSectionReader(sr, int64(t.Offset), int64(t.Size())), buf)
-				if err != nil || n != len(buf) {
-					return errors.New("read failed")
+				for _, tt := range tts {
+					func() {
```

Comment:
It might be simpler without the closure:

```
cbytes := C.CBytes(bts[:n])
C.ggml_backend_tensor_set(tt, cbytes, C.size_t(s), C.size_t(n))
defer C.free(cbytes)
```